### ISSUE TITLE
Fastify 5.8.2 and autohooks produces "route already declared" error

### ISSUE THREAD

**Issue by mccare** [fastify/fastify-autoload#376]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.27.0

### Plugin version

5.8.2

### Node.js version

20.13.1

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

14.5

### Description

When going from 5.8.0 to 5.8.2 my code base will throw a `FastifyError [Error]: Method 'GET' already declared for route '/backend'` error about having a duplicate route. I managed to reproduce this and added a simple repository to show the error.

### Link to code that reproduces the bug

https://github.com/mccare/fastify-autoload-hook

### Expected Behavior

It should just register the route into /backend/example (in this case) and not try to register under /backend

**Comment by climba03003** [comment]
After double check the document, I can see it as intended behavior.
When `options.prefix` is provided, it is used for all `routes` found. 
That means all the routes plugin is prefixed by the same prefix.

Providing `/` route inside should throw an Error.
I don't knows why it works in previous version, maybe different combination of options provide different behavior (wrong behavior).

**Comment by mccare** [comment]
So adding the prefix will just add the '/backend' to every route right? But the directory would still be appended to the prefix, so in the example:

* GET /backend
* GET /backend/example
 
Where is the "duplicate route" error coming from then?

**Comment by climba03003** [comment]
> So adding the prefix will just add the '/backend' to every route right?

No, from the document is that all routes will use the `prefix` and it will not respect the directory structure.
That is the reason why my PR unit test failed.

This happen only when you don't use `autoHooks` which means `autoHooks` is overrided the behavior or is being wrong in the beginning.

**Comment by climba03003** [comment]
I think we should decide which behavior is the correct one.
To me, it should be appending to the prefix. 

But you can see it is very weird in the test below. It ignore the folder `child`.
https://github.com/fastify/fastify-autoload/blob/c1685f4b7a971c29096dad7ed6f753d5e3040a40/test/issues/326/autohooks/index.js#L13-L18

https://github.com/fastify/fastify-autoload/blob/c1685f4b7a971c29096dad7ed6f753d5e3040a40/test/issues/326/test.js#L64-L79

cc @fastify/core

**Comment by jsumners** [comment]
@climba03003 can you summarize what question you are asking core?

**Comment by climba03003** [comment]
> can you summarize what question you are asking core?

The current `options.prefix` is not clear enough what is the expected behavior.
It can be passed to all found `routes` with that prefix but NOT extends the directory name.
or
It can be passed to all found `routes` with that prefix AND extends the directory name.

I prefer the latter.

![image](https://github.com/fastify/fastify-autoload/assets/23028015/409bffd8-3fd4-4706-8a14-a255fc5754fe)

**Comment by jsumners** [comment]
I really don't have an opinion. I have not used this module.

**Comment by mccare** [comment]
For me, just removing the autohooks.js will fix my problem. I can live with the missing encapsulation for now. Feel free to close.

**Comment by paul-norman** [comment]
I've hit this problem too, it broke my system when moving from `5.8.0` to `5.8.1`. I've reverted to `5.8.0` for the time being.

I'm using it in what seemed like a logical manner to me:

```js
fastify.register(AutoLoad, {
    dir:          'routes/api',
    options:      { prefix: '/api' },
    autoHooks:    true,
    cascadeHooks: true
});
```

and then a `routes/api` folder structure similar to:

  - routes.js *(containing: `/` and `/test`)*
  - autohooks.js
  - [subfolder1]
    - routes.js *(containing: `/` and `/sub1test`)*
    - autohooks.js
  - [subfolder2]
    - routes.js *(containing: `/` and `/sub2test`)*
    - autohooks.js

I would expect the following routes to exist:

```
// Using autohooks from "routes/api"
/api/
/api/test

// Using autohooks from "routes/api" and "routes/api/subfolder1"
/api/subfolder1/
/api/subfolder1/sub1test

// Using autohooks from "routes/api" and "routes/api/subfolder2"
/api/subfolder2/
/api/subfolder2/sub2test
```

If this will not be the case going forward, how do I achieve this goal in the newer versions?
